### DISCUSSION TITLE
OneDrive on a laptop

### DISCUSSION THREAD

**Discussion by ericc-dream** [abraunegg/onedrive#2077] [Q&A] [answered]
Hi 
I have an issue, and I couldn't find the correct information anywhere, but I can't imagine that I'm the only one in this situation.
So I have OneDrive installed on my laptop running Ubuntu 20.04 LTS , with the last version of OneDrive
It's running automatically thanks to "systemctl --user" 

When I'm in the office, directly connected to the company network, everything is OK and works as expected.
When I'm working from home connected through the VPN, OneDrive is unable to connect and I must, or stopping it, or re-authorise it.
If I re-authorise, it works but the next time I go to the office, I must re-authorise it again 
If I'm stopping it, the next time I go to the office, it start automatically, and works without complain (no need to re-authorise)

Is there a way to avoid this ? 

This is the content of syslog if it's needed 
```
 Aug  5 09:25:12 ubuntu systemd[5566]: Started OneDrive Free Client.
Aug  5 09:25:12 ubuntu systemd[139947]: onedrive.service: ProtectHostname=yes is configured, but UTS namespace setup is prohibited (container manager?), ignoring namespace setup.
Aug  5 09:25:14 ubuntu onedrive[139947]: Configuring Global Azure AD Endpoints
Aug  5 09:25:15 ubuntu onedrive[139947]: ERROR: Microsoft OneDrive API returned an error with the following message:
Aug  5 09:25:15 ubuntu onedrive[139947]:   Error Message:    HTTP request returned status code 400 (Bad Request)
Aug  5 09:25:15 ubuntu onedrive[139947]:   Error Reason:     AADSTS50076: Due to a configuration change made by your administrator, or because you moved to a new location, you must use multi-factor authentication to access '00000003-0000-0000-c000-000000000000'.
Aug  5 09:25:15 ubuntu onedrive[139947]: Trace ID: aa2634b0-a369-4cd8-be06-a46e35826300
Aug  5 09:25:15 ubuntu onedrive[139947]: Correlation ID: e7b70c85-9d52-497e-b3b2-71b06d7a2c24
Aug  5 09:25:15 ubuntu onedrive[139947]: Timestamp: 2022-08-05 07:25:15Z
Aug  5 09:25:16 ubuntu onedrive[139947]: ERROR: Microsoft OneDrive API returned an error with the following message:
Aug  5 09:25:16 ubuntu onedrive[139947]:   Error Message:    HTTP request returned status code 401 (Unauthorized)
Aug  5 09:25:16 ubuntu onedrive[139947]:   Error Reason:     Access token is empty.
Aug  5 09:25:16 ubuntu onedrive[139947]:   Error Timestamp:  2022-08-05T07:25:16
Aug  5 09:25:16 ubuntu onedrive[139947]:   API Request ID:   10ea3d25-973a-4cdc-a2e2-5822b4036bce
Aug  5 09:25:16 ubuntu onedrive[139947]: ERROR: Check your configuration as your refresh_token may be empty or invalid. You may need to issue a --reauth and re-authorise this client.
Aug  5 09:25:16 ubuntu systemd[5566]: onedrive.service: Main process exited, code=exited, status=255/EXCEPTION
Aug  5 09:25:16 ubuntu systemd[5566]: onedrive.service: Failed with result 'exit-code'.
```

Thanks in advance for your help

**Comment by abraunegg** [comment, marked as answer]
@ericc-dream 
Interesting issue you have there .. however, unfortunately the 'onedrive' client here is not at fault - it is operating 100% normally.

> So I have OneDrive installed on my laptop running Ubuntu 20.04 LTS , with the last version of OneDrive

Because you are using Ubuntu ... please make sure you are running v2.4.20 of the client and you installed the client following this process: https://github.com/abraunegg/onedrive/blob/master/docs/ubuntu-package-install.md

> Is there a way to avoid this ?

Without understanding more of your actual environment at home (or any other location that needs to you use a VPN) - I would start by having a look at your VPN policies / VPN configuration dealing with VPN split tunneling.

From your description, it 'sounds' like when you are using your VPN, your traffic that you would expect to go via the the VPN is actually going via the Internet - not via your VPN tunnel - which, is why you are being requested to re-auth .. because your prior authenticated state was from an approved network - and now your coming over the Internet (unapproved network).

Unfortunately debugging this | helping you get to the bottom of this is not something I can really assist with here other than the following suggestions:
* When in the office, use Wireshark to capture the authentication process, note the URLs and IP addresses the client is accessing during the authentication process.
* When working via VPN, do the same - capture the authentication process with Wireshark
* Compare the Wireshark results, then compare this to your VPN configuration - do you have your VPN split-tunnel configuration done right so that access of those URL's / OneDrive servers go via the VPN and not out via your normal Internet connection
* Understand what route your traffic is taking when using the VPN ... use `route` to understand what default gateway traffic to OneDrive is taking
* Speak to your network & security folk at work to also assist you - maybe they missed a network subnet on their end - so that VPN users are mis-configured - as maybe your VPN configuration is correct .... 
* Reconfigure your VPN so that your OneDrive traffic goes via the VPN

**Reply by ericc-dream** [reply]
I have the last version on OneDrive installed and configured 
> ~$ onedrive --version
> onedrive v2.4.20-1+np1

> From your description, it 'sounds' like when you are using your VPN, your traffic that you would expect to go via the the VPN is actually going via the Internet - not via your VPN tunnel - which, is why you are being requested to re-auth .. because your prior authenticated state was from an approved network - and now your coming over the Internet (unapproved network).

This was the missing information. 
Indeed I had a script to redirect the traffic to not pass through the VPN when it wasn't absolutely needed ! 
I removed the script and reconnect to the VPN, and now everything is working as expected ... 
I wasn't aware of this "approved/unapproved" network

Thanks

**Comment by abraunegg** [comment]
@ericc-dream 
Additionally, googling 'AADSTS50076' turns up this piece of advice by Microsoft:

https://docs.microsoft.com/en-us/answers/questions/836429/aadsts50076-due-to-a-configuration-change-made-by.html

Essentially:
* Create a new application ID
* Configure the application to use application permissions rather that delegated permissions - refer to https://github.com/abraunegg/onedrive/blob/master/docs/national-cloud-deployments.md on how to do this

So potentially to resolve this, you will need to speak to your network & security folk at work and whoever looks after your Microsoft Azure Directory
